### ISSUE TITLE
JDK23 jdk_lang_1_FAILED java/lang/String/CompactString/MaxSizeUTF16String.java OutOfMemoryError: Required length exceeds implementation limit

### ISSUE THREAD

**Issue by JasonFengJ9** [eclipse-openj9/openj9#20035]
Failure link
------------

From [an internal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk23_j9_sanity.openjdk_aarch64_mac/12/)(`macaarch64rt1`):
```
openjdk version "23-beta" 2024-09-17
IBM Semeru Runtime Open Edition 23+36-202408210002 (build 23-beta+36-202408210002)
Eclipse OpenJ9 VM 23+36-202408210002 (build master-a308eaf9e0, JRE 23 Mac OS X aarch64-64-Bit 20240821_18 (JIT enabled, AOT enabled)
OpenJ9   - a308eaf9e0
OMR      - e0dde2287
JCL      - f97359bdc based on jdk-23+36)
```
[Rerun in Grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/parambuild/?SDK_RESOURCE=customized&amp;TARGET=jdk_lang_1&amp;BASE_DOCKER_REGISTRY_CREDENTIAL_ID=&amp;TEST_FLAG=&amp;UPSTREAM_TEST_JOB_NAME=&amp;DOCKER_REQUIRED=false&amp;ACTIVE_NODE_TIMEOUT=5&amp;VENDOR_TEST_DIRS=&amp;EXTRA_DOCKER_ARGS=&amp;TKG_OWNER_BRANCH=adoptium%3Amaster&amp;OPENJ9_SYSTEMTEST_OWNER_BRANCH=eclipse%3Amaster&amp;PLATFORM=aarch64_mac&amp;GENERATE_JOBS=true&amp;KEEP_REPORTDIR=true&amp;PERSONAL_BUILD=false&amp;DOCKER_REGISTRY_DIR=&amp;RERUN_ITERATIONS=0&amp;ADOPTOPENJDK_REPO=https%3A%2F%2Fgithub.com%2Fadoptium%2Faqa-tests.git&amp;SETUP_JCK_RUN=false&amp;DOCKER_REGISTRY_URL_CREDENTIAL_ID=&amp;LABEL=&amp;EXTRA_OPTIONS=&amp;CUSTOMIZED_SDK_URL=https%3A%2F%2Fna-public.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk23%2Fjdk23-mac-aarch64-openj9%2F18%2Fibm-semeru-open-jdk_aarch64_mac_JDK23U_2024-08-21-00-00.tar.gz+https%3A%2F%2Fna-public.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk23%2Fjdk23-mac-aarch64-openj9%2F18%2Fibm-semeru-open-testimage_aarch64_mac_JDK23U_2024-08-21-00-00.tar.gz&amp;BUILD_IDENTIFIER=&amp;JENKINS_KEY=Jenkins+4096+key&amp;ADOPTOPENJDK_BRANCH=master&amp;LIGHT_WEIGHT_CHECKOUT=false&amp;USE_JRE=false&amp;ARTIFACTORY_SERVER=na.artifactory.swg-devops&amp;KEEP_WORKSPACE=false&amp;USER_CREDENTIALS_ID=83181e25-eea4-4f55-8b3e-e79615733226&amp;JDK_VERSION=23&amp;DOCKER_REGISTRY_URL=&amp;ITERATIONS=1&amp;VENDOR_TEST_REPOS=&amp;JDK_REPO=git%40github.com%3Aibmruntimes%2Fopenj9-openjdk-jdk23&amp;JCK_GIT_BRANCH=main&amp;RELEASE_TAG=&amp;OPENJ9_BRANCH=master&amp;OPENJ9_SHA=&amp;JCK_GIT_REPO=&amp;VENDOR_TEST_BRANCHES=&amp;OPENJ9_REPO=https%3A%2F%2Fgithub.com%2Feclipse-openj9%2Fopenj9.git&amp;UPSTREAM_JOB_NAME=&amp;CLOUD_PROVIDER=&amp;CUSTOM_TARGET=&amp;VENDOR_TEST_SHAS=&amp;JDK_BRANCH=openj9&amp;LABEL_ADDITION=&amp;ARTIFACTORY_REPO=sys-rt-generic-local%2Chyc-runtimes-team-restricted-oracle-sdk-generic-local&amp;ARTIFACTORY_ROOT_DIR=&amp;UPSTREAM_TEST_JOB_NUMBER=&amp;DOCKERIMAGE_TAG=&amp;TEST_TIME=120&amp;JDK_IMPL=openj9&amp;SSH_AGENT_CREDENTIAL=83181e25-eea4-4f55-8b3e-e79615733226&amp;AUTO_DETECT=true&amp;SLACK_CHANNEL=%23rt-jenkins&amp;DYNAMIC_COMPILE=false&amp;RELATED_NODES=&amp;ADOPTOPENJDK_SYSTEMTEST_OWNER_BRANCH=adoptium%3Amaster&amp;APPLICATION_OPTIONS=&amp;CUSTOMIZED_SDK_URL_CREDENTIAL_ID=7c1c2c28-650f-49e0-afd1-ca6b60479546&amp;ARCHIVE_TEST_RESULTS=false&amp;NUM_MACHINES=&amp;OPENJDK_SHA=&amp;TRSS_URL=https%3A%2F%2Ftrssrtp1.fyre.ibm.com%2F&amp;USE_TESTENV_PROPERTIES=false&amp;BUILD_LIST=openjdk&amp;ADDITIONAL_ARTIFACTS_REQUIRED=&amp;UPSTREAM_JOB_NUMBER=&amp;STF_OWNER_BRANCH=adoptium%3Amaster&amp;TIME_LIMIT=25&amp;JVM_OPTIONS=&amp;PARALLEL=Dynamic) - Change TARGET to run only the failed test targets.

Optional info
-------------

Failure output (captured from console output)
---------------------------------------------
```
[2024-08-21T02:51:37.678Z] variation: -Xdump:system:none -Xdump:heap:none -Xdump:system:events=gpf+abort+traceassert+corruptcache -XX:-JITServerTechPreviewMessage Mode650
[2024-08-21T02:51:37.678Z] JVM_OPTIONS:  -Xdump:system:none -Xdump:heap:none -Xdump:system:events=gpf+abort+traceassert+corruptcache -XX:-JITServerTechPreviewMessage -XX:-UseCompressedOops -Xverbosegclog 

[2024-08-21T02:55:17.213Z] TEST: java/lang/String/CompactString/MaxSizeUTF16String.java

[2024-08-21T02:55:17.213Z] STARTED    MaxSizeUTF16String::testMaxUTF8_UTF16Encode 'testMaxUTF8_UTF16Encode()'
[2024-08-21T02:55:17.213Z] java.lang.OutOfMemoryError: Required length exceeds implementation limit
[2024-08-21T02:55:17.213Z] 	at java.base/java.lang.String.encodeUTF8_UTF16(String.java:1542)
[2024-08-21T02:55:17.213Z] 	at java.base/java.lang.String.encodeUTF8(String.java:1511)
[2024-08-21T02:55:17.213Z] 	at java.base/java.lang.String.encode(String.java:1079)
[2024-08-21T02:55:17.213Z] 	at java.base/java.lang.String.getBytes(String.java:2555)
[2024-08-21T02:55:17.213Z] 	at MaxSizeUTF16String.lambda$testMaxUTF8_UTF16Encode$0(MaxSizeUTF16String.java:153)
[2024-08-21T02:55:17.213Z] 	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:53)
[2024-08-21T02:55:17.213Z] 	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
[2024-08-21T02:55:17.213Z] 	at org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3083)
[2024-08-21T02:55:17.213Z] 	at MaxSizeUTF16String.testMaxUTF8_UTF16Encode(MaxSizeUTF16String.java:153)

[2024-08-21T02:55:17.217Z] TEST RESULT: Failed. Execution failed: `main' threw exception: java.lang.Exception: JUnit test failure
[2024-08-21T02:55:17.217Z] --------------------------------------------------
[2024-08-21T03:00:07.938Z] Test results: passed: 930; failed: 1
[2024-08-21T03:00:09.218Z] Report written to /Users/jenkins/workspace/Test_openjdk23_j9_sanity.openjdk_aarch64_mac/aqa-tests/TKG/output_17242086962796/jdk_lang_1/report/html/report.html
[2024-08-21T03:00:09.218Z] Results written to /Users/jenkins/workspace/Test_openjdk23_j9_sanity.openjdk_aarch64_mac/aqa-tests/TKG/output_17242086962796/jdk_lang_1/work
[2024-08-21T03:00:09.218Z] Error: Some tests failed or other problems occurred.
[2024-08-21T03:00:09.218Z] -----------------------------------
[2024-08-21T03:00:09.218Z] jdk_lang_1_FAILED
```

[50x internal grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/42814/) - all failed

**Comment by knn-k** [comment]
The same testcase (MaxSizeUTF16String) with different JVM option (-XX:-CompactStrings), different error: #20021

**Comment by hzongaro** [comment]
Does the above capture the right failure output?  I think the actual failure in the [console log for that internal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk23_j9_sanity.openjdk_aarch64_mac/12/consoleFull) is for test case `MaxSizeUTF16String::testMaxCharArray`:

```
22:55:17  STARTED    MaxSizeUTF16String::testMaxCharArray 'testMaxCharArray()'
22:55:17  Checking max UTF16 string len: 1073741824
22:55:17  Checking max UTF16 string len: 1073741823
22:55:17  org.opentest4j.AssertionFailedError: Expected OutOfMemoryError with message prefix: UTF16 String size is
22:55:17  	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:38)
22:55:17  	at org.junit.jupiter.api.Assertions.fail(Assertions.java:135)
22:55:17  	at MaxSizeUTF16String.testMaxCharArray(MaxSizeUTF16String.java:112)
22:55:17  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
22:55:17  	at java.base/java.lang.reflect.Method.invoke(Method.java:586)
22:55:17  	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
...
22:55:17  	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
22:55:17  	at java.base/java.lang.Thread.run(Thread.java:1587)
22:55:17  FAILED     MaxSizeUTF16String::testMaxCharArray 'testMaxCharArray()'
22:55:17  STARTED    MaxSizeUTF16String::testMaxUTF8_UTF16Encode 'testMaxUTF8_UTF16Encode()'
```

This might be a duplicate of #20021

**Comment by hzongaro** [comment]
I ran a [50x internal grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/42893/) using the fix for #20021 from pull request #20058.  It passed 100%, so I'm going to mark this as a duplicate.

**Comment by hzongaro** [comment]
Duplicate of #20021

**Comment by hzongaro** [comment]
If the problem can still be reproduced after #20058 is merged, please reopen this.

**Comment by hzongaro** [comment]
If the problem can still be reproduced after #20058 is merged, please reopen this.
